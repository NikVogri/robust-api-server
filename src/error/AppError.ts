interface IAppError {
    message: string;
    statusCode: number;
    thrownError?: Error | unknown;
    payload?: Record<string, unknown>;
}

export class AppError extends Error {
    thrownError?: Error | unknown;
    statusCode: number;
    payload?: Record<string, unknown>;

    constructor(public data: IAppError) {
        super(data.message);

        this.thrownError = data.thrownError;
        this.statusCode = data.statusCode;
        this.payload = data.payload;
    }

    static toStringifiableObject(error: Error | unknown) {
        if (!(error instanceof Error)) return {};

        const err: Record<string, unknown> = {};

        Object.getOwnPropertyNames(error).forEach((prop) => {
            if ((error[prop as keyof Error] as unknown) instanceof Error) {
                err[prop] = AppError.toStringifiableObject(error);
            } else err[prop] = error[prop as keyof Error];
        });

        return err;
    }
}
