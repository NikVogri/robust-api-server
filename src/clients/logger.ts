import { Request, Response } from 'express';
import { Service } from 'typedi';
import winston from 'winston';
import { AppError } from '../error/AppError';

@Service()
export class Logger {
    private logger: winston.Logger;

    private custom = {
        levels: {
            error: 0,
            info: 2,
            http: 3,
        },
        colors: {
            error: 'red',
            info: 'yellow',
            http: 'gray',
        },
    };

    constructor() {
        winston.addColors(this.custom.colors);
        winston.remove(winston.transports.Console);

        this.logger = winston.createLogger({
            transports: this.getTransports(),
            levels: this.custom.levels,
        });
    }

    private getTransports = (): winston.transports.ConsoleTransportInstance[] => {
        return [
            // Logs appear in STDOUT
            new winston.transports.Console({
                level: 'http',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'info',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'error',
                format: winston.format.combine(
                    // winston.format.timestamp(), TODO: add timestamp
                    // TODO: make sure error only gets logged once! remove default transports for error?
                    winston.format.json(),
                    winston.format.colorize({ all: true }),
                ),
            }),
        ];
    };

    error(message: string, error: Error, metadata?: Record<string, unknown>) {
        this.logger.error(message, AppError.toStringifiableObject(error), metadata);
    }

    info(message: string, metadata?: Record<string, unknown>) {
        this.logger.info(message, metadata);
    }

    http(req: Request, res: Response, error?: AppError) {
        const log = this.buildHttpLog(req, res, error);
        this.logger.http(log);
    }

    private buildHttpLog(req: Request, res: Response, error?: AppError): string {
        return JSON.stringify({
            success: res.statusCode.toString().startsWith('2'), // success status codes are 2xx
            url: req.url,
            originalUrl: req.originalUrl,
            path: `${req.method} ${req.url}`,
            // TODO: Can also add auth fields in the future
            // user: {}
            client: {
                ip: req.ip,
                hostname: req.hostname,
                version: req.httpVersion,
                userAgent: req.headers['user-agent'],
            },
            req: {
                query: req.query,
                body: req.body ? Object.keys(req.body) : undefined,
                // do not log request inputs as they might be too large or contain GDPR data
            },
            res: {
                statusCode: res.statusCode,
                // do not log resposes as they might be too large or contain GDPR data
            },
            error: {
                message: error?.message,
                stack: error?.stack,
                thrown: error?.thrownError ? AppError.toStringifiableObject(error.thrownError) : undefined,
            },
        });
    }
}
