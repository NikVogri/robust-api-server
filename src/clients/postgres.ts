import { Pool } from 'pg';
import { Logger } from './logger';
import { singleton } from 'tsyringe';

@singleton()
export class Postgres {
    readonly pool: Pool;

    constructor(private logger: Logger) {
        this.pool = new Pool({
            connectionString: process.env.PG_DB_CONNECTION_STR,
        });
    }

    async testConnection(): Promise<void> {
        await this.pool.query('SELECT 1 + 1;');
        this.logger.info('Successfully connected to Postgres');
    }

    async getClient() {
        return await this.pool.connect();
    }
}
