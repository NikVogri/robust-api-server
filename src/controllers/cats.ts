import { Service } from 'typedi';
import { CatsService } from '../services/cats';
import { Request, Response } from 'express';

@Service()
export class CatsController {
    constructor(private catsService: CatsService) {}

    async getAll(_: Request, res: Response) {
        const cats = await this.catsService.getAll();
        res.status(200).send(cats);
    }

    async getById(req: Request, res: Response) {
        const id = req.params.id;
        const cat = await this.catsService.getById(id);

        res.status(200).send(cat);
    }
}
