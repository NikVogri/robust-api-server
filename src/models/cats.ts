import { Model } from './base';
import { Cat } from './types';
import { Postgres } from '../clients/postgres';
import { injectable } from 'tsyringe';
import { CreateCatDto } from '../routes/cats.dto';

@injectable()
export class CatsModel extends Model<Cat> {
    constructor(db: Postgres) {
        super('cats', db);
    }

    async create(createCatDto: CreateCatDto): Promise<Cat> {
        const { rows } = await this.db.pool.query(
            `
                INSERT INTO cats (name, date_of_birth, color)
                VALUES ($1, $2, $3)
                RETURNING *;
                `,
            [createCatDto.name, createCatDto.dateOfBirth, createCatDto.color],
        );

        return rows[0];
    }
}
