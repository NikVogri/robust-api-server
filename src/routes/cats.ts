import express from 'express';
import { withErrorHandling } from '../middleware/withCatch';
import Container from 'typedi';
import { CatsController } from '../controllers/cats';

const catsRouter = express.Router();
const catsController = Container.get(CatsController);

catsRouter.get(
    '/cats',
    withErrorHandling((req, res) => catsController.getAll(req, res)),
);

catsRouter.get(
    '/cats/:id',
    withErrorHandling((req, res) => catsController.getById(req, res)),
);

export { catsRouter };
