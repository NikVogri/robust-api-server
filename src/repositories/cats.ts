import { Repository } from './repository';
import { CatDBItem } from '../models';
import { Postgres } from '../services/database/postgres';
import { injectable } from 'tsyringe';
import { CreateCatDto } from '../routes/cats.dto';
import { AppError } from '../error/AppError';

@injectable()
export class CatsRepository extends Repository<CatDBItem> {
    constructor(pg: Postgres) {
        super('cats', pg);
    }

    async create(createCatDto: CreateCatDto): Promise<CatDBItem> {
        try {
            const { rows } = await this.pg.pool.query(
                `
                INSERT INTO cats (name, date_of_birth, color)
                VALUES ($1, $2, $3)
                RETURNING *;
                `,
                [createCatDto.name, createCatDto.dateOfBirth, createCatDto.color],
            );

            return rows[0];
        } catch (err) {
            throw new AppError({
                message: 'Unable to create a new cat',
                statusCode: 500,
                thrownError: err,
            });
        }
    }
}
