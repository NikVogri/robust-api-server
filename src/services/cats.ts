import { CatDBItem } from '../models';
import { AppError } from '../error/AppError';
import { Bull } from '../clients/bull';
import { CatsRepository } from '../repositories/cats';
import { injectable } from 'tsyringe';
import { CreateCatDto } from '../routes/cats.dto';

@injectable()
export class CatsService {
    constructor(
        private catsRepo: CatsRepository,
        private bull: Bull,
    ) {}

    async getAll(): Promise<CatDBItem[]> {
        const cats = await this.catsRepo.getAll();
        return cats;
    }

    async getById(id: string): Promise<CatDBItem> {
        const cat = await this.catsRepo.getById(id);

        if (!cat) {
            throw new AppError({
                message: 'Cat does not exist :(',
                statusCode: 404,
            });
        }

        // don't forget to pet the cat
        await this.bull.queues.catPetterQueue.add('pet-cat', { name: cat.name });

        return cat;
    }

    async createCat(createCatDto: CreateCatDto): Promise<CatDBItem> {
        const cat = await this.catsRepo.create(createCatDto);
        return cat;
    }
}
