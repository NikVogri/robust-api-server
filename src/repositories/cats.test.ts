import { describe, expect, it, vi } from 'vitest';
import { Postgres } from '../services/database/postgres';
import { Logger } from '../services/logging/logger';
import { CatsRepository } from './cats';
import { AppError } from '../error/AppError';

vi.mock('../services/logging/logger');
vi.mock('../services/database/postgres', () => {
    return {
        Postgres: vi.fn(() => ({
            pool: {
                query: vi.fn(),
            },
        })),
    };
});

describe('Test CatsRepository', () => {
    const logger = new Logger();
    const pg = new Postgres(logger);

    const catsRepo = new CatsRepository(pg);

    it('should create a new cat record', async () => {
        expect.assertions(2);

        vi.spyOn(pg.pool, 'query').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            return Promise.resolve({
                rows: [{ id: 1 }],
            });
        });

        const res = await catsRepo.create({ name: 'Kiki', dateOfBirth: '2010-09-05', color: 'gray' });
        expect(res).toMatchSnapshot();
    });

    it('should throw an app error when unable to create a cat', async () => {
        expect.assertions(2);

        vi.spyOn(pg.pool, 'query').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            throw new Error('something went wrong');
        });

        try {
            await catsRepo.create({ name: 'Kiki', dateOfBirth: '2010-09-05', color: 'gray' });
        } catch (err) {
            expect(err).toBeInstanceOf(AppError);
        }
    });
});
