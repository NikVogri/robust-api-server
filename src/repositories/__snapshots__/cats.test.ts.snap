// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test CatsRepository > should create a new cat record 1`] = `
[
  "
                INSERT INTO cats (name, date_of_birth, color)
                VALUES ($1, $2, $3)
                RETURNING *;
                ",
  [
    "Kiki",
    "2010-09-05",
    "gray",
  ],
]
`;

exports[`Test CatsRepository > should create a new cat record 2`] = `
{
  "id": 1,
}
`;

exports[`Test CatsRepository > should throw an app error when unable to create a cat 1`] = `
[
  "
                INSERT INTO cats (name, date_of_birth, color)
                VALUES ($1, $2, $3)
                RETURNING *;
                ",
  [
    "Kiki",
    "2010-09-05",
    "gray",
  ],
]
`;
